src/dtx/dtx_common.c

```diff
@@ -565,77 +581,37 @@ dtx_leader_end(struct dtx_leader_handle *dlh, struct ds_cont_hdl *cont_hdl,
 		if (rc >= 0) {
 			D_DEBUG(DB_TRACE, "retry DTX "DF_DTI"\n",
 				DP_DTI(&dth->dth_xid));
-			D_GOTO(out, result = -DER_AGAIN);
+			return -DER_AGAIN;
 		}
-
-		if (result == 0)
-			result = rc;
-
-		D_GOTO(fail, result);
-	} else if (rc < 0 || daos_is_zero_dti(&dth->dth_xid)) {
-		if (result == 0)
-			result = rc;
-
-		D_GOTO(fail, result);
 	}
 
-	if (result < 0 || rc < 0 || daos_is_zero_dti(&dth->dth_xid))
-		D_GOTO(fail, result = (result == 0 ? rc : result));
-
-	/* If the DTX is started befoe DTX resync operation (for rebuild),
-	 * then it is possbile that the DTX resync ULT may have aborted
-	 * or committed the DTX during current ULT waiting for the reply.
-	 * let's check DTX status locally before marking as 'committable'.
-	 */
-	if (dlh->dlh_handled_time <= cont->sc_dtx_resync_time) {
-		rc = vos_dtx_check(cont->sc_hdl, &dth->dth_xid);
-		switch (rc) {
-		case DTX_ST_PREPARED:
-			rc = vos_dtx_lookup_cos(dth->dth_coh, &dth->dth_oid,
-					&dth->dth_xid, dth->dth_dkey_hash,
-					dth->dth_intent == DAOS_INTENT_PUNCH ?
-					true : false);
-			/* The resync ULT has already added it into the CoS
-			 * cache, current ULT needs to do nothing.
-			 */
-			if (rc == 0)
-				D_GOTO(out_free, result = 0);
-
-			/* normal case, then add it to CoS cache. */
-			if (rc == -DER_NONEXIST)
-				break;
-
-			D_GOTO(fail, result = (rc >= 0 ? -DER_INVAL : rc));
-		case DTX_ST_COMMITTED:
-			/* The DTX has been committed by resync ULT by race,
-			 * set dth_sync to indicate that in log.
-			 */
-			dth->dth_sync = 1;
-			D_GOTO(out_free, result = 0);
-		case -DER_NONEXIST:
-			/* The DTX has been aborted by resync ULT, ask the
-			 * client to retry via returning -DER_INPROGRESS.
-			 */
-			result = -DER_INPROGRESS;
-			goto out_free;
-		default:
-			result = rc >= 0 ? -DER_INVAL : rc;
-			goto fail;
-		}
-	}
+	if (result < 0 || rc < 0)
+		D_GOTO(out, result = result < 0 ? result : rc);
 
 	rc = vos_dtx_add_cos(dth->dth_coh, &dth->dth_oid, &dth->dth_xid,
-			     dth->dth_dkey_hash, dth->dth_epoch,
+			     dth->dth_dkey_hash, dth->dth_epoch, dth->dth_gen,
 			     dth->dth_intent == DAOS_INTENT_PUNCH ?
-			     true : false, true);
+			     true : false);
 	if (rc == -DER_INPROGRESS) {
 		D_WARN(DF_UUID": Fail to add DTX "DF_DTI" to CoS "
 		       "because of using old epoch "DF_U64"\n",
 		       DP_UUID(cont->sc_uuid), DP_DTI(&dth->dth_xid),
 		       dth->dth_epoch);
-		D_GOTO(fail, result = rc);
+		D_GOTO(out, result = rc);
 	}
 
+	/* The DTX has been aborted by resync ULT, ask the client to retry. */
+	if (rc == -DER_NONEXIST) {
+		D_WARN(DF_UUID": Fail to add DTX "DF_DTI" with eph "DF_U64
+		       "to CoS because of being aborted when resync by race\n",
```

Comment:
missing space here so epoch will be mashed with to COS

Reply:
Done in the latest version.